### PR TITLE
ENH: plot_density for KDE plotting of point patterns based on statsmodels

### PR DESCRIPTION
**Pull request by martinfleis**
Resolves https://github.com/gdsbook/book/issues/305. Essentially replaces `seaborn.kdeplot` which does not work as expected with our own solution.

Tests will come once I'll figure out how to do them.

```py
import geopandas as gpd
from pointpats.kde import plot_density

gdf = gpd.read_file('https://martinfleischmann.net/sds/chapter_06/data/brno_pedestrian_accidents.gpkg')


gdf.plot(markersize=0.5, figsize=(9,9))
plot_density(gdf, 500)
```
4d2bc015-5f66-441c-8b69-c488ecd58d2c

```py
gdf.plot(markersize=0.5, figsize=(9,9))
plot_density(gdf, 500, levels=25, cmap="magma_r", linewidths=1)
```

60e0adc4-0c3b-4f59-a7f8-1a7d07c824a8

```py
gdf.plot(markersize=0.1, figsize=(9,9), zorder=2, color='k')
plot_density(gdf, 500, levels=25, cmap="magma_r", fill=True)
```
d68543e7-cbcb-458d-872e-ce62ea3f3da1

```py
gdf.plot(markersize=0.5, figsize=(9,9))
plot_density(gdf, 1000, cmap="magma_r", linewidths=1)
```
ceb74c5e-0c3c-4f63-ad6b-b4d2f127b72a

I would love to have this in the next release alongside #111 so I can use both during teaching starting in two weeks.

### CODE REVIEW

**Comment by knaaptime** [comment]
funny. I'm also right in the middle of using this for some bike/ped accident work

File: pointpats/kde.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import numpy as np
+
+
+def plot_density(
+    data, bandwidth, resolution=100, levels=10, fill=False, margin=0.1, **kwargs
+):
+    """Plot kernel density of a given point pattern
+
+    This uses s``tatsmodels.nonparametric.KDEMultivariate`` class to create
```
Comment (knaaptime) [inline pointpats/kde.py:9]: splice on the quote here

**Comment by martinfleis** [comment]
I was struggling to make it work decently fast on large data so I included also an implementation on top of `KDEpy`. That is fast, offers a selection of kernels but can behave funny in some edge cases. While statsmodels seems to be robust default.

**Comment by martinfleis** [comment]
The tests are not perfect but testing matplotlib is just a pain in the ass, so this at least ensures that the stuff works and args are used. It should be ready for review.

File: pointpats/kde.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    levels : int or array-like, optional
+        Determines the number and positions of the contour lines / regions.
+        See the documentation of ``matplotlib.pyplot.contour`` for details., by default 10
+    fill : bool, optional
+        Fill the area between contour lines, by default False
+    margin : float, optional
+        The factor of the margin by which the extent of the data will be expanded when
+        creating the grid. 0.1 means 10% on each side, by default 0.1. Only used
+        with the ``statsmodels`` implementation.
+
```
Comment (jGaboardi) [inline pointpats/kde.py:55]: Should we a note here that the `kwargs` here are supplied to `matplotlib.pyplot.contourf`?

Comment (martinfleis) [reply pointpats/kde.py:55]: Ah, yes. One sec

**Review by jGaboardi — APPROVED** [review]

**Review by sjsrey — APPROVED** [review]

**Comment by knaaptime** [comment]
right now this gives back a `matplotlib.contour.QuadContourSet`. Is there any way to have it return an axes object, e.g. to play nicely with contextily?
